### PR TITLE
Optimize swap gates

### PR DESCRIPTION
**Pull request by daxfohl**
This small PR optimizes handling of pure swap gates by swapping indexes instead of applying a unitary. It also avoids merging distinct state spaces when swapping qubits across them.

Related #3240
Possibly addresses #1060

### CODE REVIEW

File: cirq-core/cirq/sim/act_on_args.py
Comment (daxfohl) [inline cirq-core/cirq/sim/act_on_args.py:125]: Validate equal dimension.

File: cirq-core/cirq/sim/act_on_args.py
Comment (daxfohl) [inline cirq-core/cirq/sim/act_on_args.py:136]: Validate equal dimension.

**Review by daxfohl — COMMENTED** [review]
Validate equal dimension.

**Comment by daxfohl** [comment]
Leaning toward leaving it as-is (functional vs inplace). Other meta-ops on ActOnArgs are functional. I think consistency is important.

We can add a swap_inplace and rename_inplace, etc. if we eventually want to, as well as reorder_inplace, extract_inplace, etc. In fact, a swap_inplace would be useful on OperationTarget, and then this optimization could be called directly in _core_iterator, thus working even when split_untangled_qubits is false.

File: cirq-core/cirq/linalg/transformations.py
Comment (daxfohl) [inline cirq-core/cirq/linalg/transformations.py:510]: Fix merge issue here.

File: cirq-core/cirq/sim/act_on_args_container.py
Comment (daxfohl) [inline cirq-core/cirq/sim/act_on_args_container.py:80]: Tests

File: cirq-core/cirq/sim/act_on_args.py
Comment (daxfohl) [inline cirq-core/cirq/sim/act_on_args.py:151]: @95-martin-orion Are you okay with `copy.copy` here for a shallow copy? I don't see it used much elsewhere. Another option would be to make ActOnArgs and descendants all dataclasses and use `dataclasses.replace` here. I think it makes sense to do so if we're already using dataclasses elsewhere, though I don't have any experience with dataclasses so I might be wrong. (For instance, there's no sense in making them frozen, because the state they represent changes and would be too expensive to copy for each change, but it doesn't look like being frozen is a requirement for a dataclass).

Note a fair chunk of #4265 is also "cleaning stuff up" by using `copy.copy`, and should probably be split into two separate PRs. I just wanted to reduce the amount duplication I noticed while adding CH-form.

Comment (95-martin-orion) [reply cirq-core/cirq/sim/act_on_args.py:151]: `copy.copy` is fine for shallow copy, but is that what we want here? Sharing a `prng` reference between two `ActOnArgs` could generate non-reproducible results.

We have a few other `dataclass`es in Cirq, so I'd be fine with using it for `ActOnArgs` - although in my experience `dataclass` inheritance with default values can be awkward.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Went ahead and reviewed the rest of this PR - the "shallow vs. deep copy" is my only concern.

File: cirq-core/cirq/sim/act_on_args.py
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py:151]: It behaves like methods previously known as split, join, and reindex, which also do shallow copies (though those methods create the copy manually rather than use copy.copy -- changing those to use `copy.copy` is the "cleanup" section of #4265 I mentioned above). The prng is "owned" by the simulator itself, so they all keep a reference to that.

Comment (95-martin-orion) [reply cirq-core/cirq/sim/act_on_args.py:151]: I guess there's an implicit assumption here that the calling `ActOnArgs` object will be discarded afterwards? I think the `inplace` argument is what's confusing me - if both `inplace=True` and `inplace=False` will invalidate the old object, why should we have an `inplace=False` mode at all?

Comment (95-martin-orion) [reply cirq-core/cirq/sim/act_on_args.py:151]: As an aside: if the implicit assumption above is correct, could you update the docstrings of both this and the split/join/reindex methods to clarify this?

Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py:151]: Allowing swap to be inplace cleans up the code a lot in this PR, because otherwise we'd have to go through and backfill all the dictionaries afterward with the new reference in several places. But the existing methods (split/join/reorder) all create new copies instead of operating inplace. It seemed awkward to introduce a new semantic just for this method.

So rather than having different semantics for different methods, or to have a single semantic that may be suboptimal for some cases, I figured the best approach would be to standardize on a parameter for all these methods that lets the user choose, especially since it only requires a single line of code. (I added this parameter to the remaining functions in #4265). This is also the approach numpy itself uses in a lot of functions.

Yes I think it's good to make that explicit. I'll update the docstrings, and then reupdate them in #4265.

**Comment by 95-martin-orion** [comment]
> Went ahead and reviewed the rest of this PR - the "shallow vs. deep copy" is my only concern.

As discussed above, this is the expected behavior. Just waiting on docstring updates for approval.

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**
